Fix Circle to remove singular edge

GlennWSo opened the pull request:
### Commits
- added test assert for circle:   all edges of the circle should have approx the same length
- Remove duplicate point, by setting endpoint=False for theta This will not change the total number of points, but the spacing of theta, now passes the updated test

### Overview

<!-- Please insert a high-level description of this pull request here. -->
Fixing bugg in pv.Circle. 
See linked issue below, for a more detailed bugg description:
resolves #3709 

### Breaking Change:
This change can break functions depending on pyvista.Circle
#### Example of broken function:

The fallowing function works with current pyvista, but with the suggested changes it will be broken.
With the changes, the hexagon function will incorrectly create a 7-sided polygon.
```python
import pyvista as pv


def hexagon(r: float) -> pv.PolyData:
    """
    creates a regular hexagon(polygon with 6 equal sides) in plane: Z=0,
     with points <r> distance from origo
    """
    # create one extra point and edge, since circle always make one incorrect edge
    hex = pv.Circle(radius=r, resolution=6 + 1)
    return hex.clean(tolerance=0.001)  # clean away the edge with near zero length
```
With the suggested pyvista changes, the hexagon implementation would need too be updated like so:
 
```python
import pyvista as pv

def hexagon(r: float) -> pv.PolyData:
    """
    creates a regular hexagon(polygon with 6 equal sides) in plane: Z=0,
     with points <r> distance from origo
    """
    hex = pv.Circle(radius=r, resolution=6)
    return hex
```

### Details

- < feature1 or bug1 description >

```python
import pyvista as pv
import numpy as np

circle4 = pv.Circle(radius=1, resolution=4)
edge_vectors = np.roll(circle4.points, shift=1, axis=0)/4  - circle4.points
print(edge_vectors)
circle4.plot()
```
#### now prints:
>[[-1., -1.,  0.],
[ 1., -1.,  0.],
[ 1.,  1.,  0.],
[-1.,  1.,  0.]])

#### old prints:
>[[ 0.     -0.      0.    ]
 [ 1.5    -0.866   0.    ]
 [ 0.      1.7321  0.    ]
 [-1.5    -0.866   0.    ]]

### plots
#### new
![circle4](https://user-images.githubusercontent.com/107837123/207076452-cad8a6c2-686f-4ef6-982c-0f0cf7af39da.png)

#### old
![old](https://user-images.githubusercontent.com/107837123/207076121-f38ad095-d5d5-42c1-8743-feb88f2986e0.png)

banesullivan reviewed (approved):
Thank you for fixing this!

tkoyama010 reviewed (approved):
LGTM. Thanks for fixing this.

adeak commented:
Good catch on the issue, thanks @GlennWSo.

Just to muse a bit about the choice we have to make when fixing the bug: either keep the number of non-degenerate edges, or keep the number of actual edges. This PR chose the latter, which means that _visually_ the result will change, and so will the point coordinates, but the size of the mesh will be the same. I'd be tempted to think that opting to keep the number of non-degenerate edges (i.e. sticking with a triangle for `resolution=4`) would have the potential to break downstream code relying on the exact size of the polygon (which would arguably be a _good_ thing).

Anyway, the documentation of the helper clearly states that

>     resolution : int, default: 100
        Number of points on the circle.

So the current PR's choice is a clear bug-fix that makes the function behave according to the API. Another question is whether this merits a [breaking-change] label... I'm only wondering about this because of this will introduce subtle changes with no warning to end users.

GlennWSo commented:
> Good catch on the issue, thanks @GlennWSo.
> 
> Just to muse a bit about the choice we have to make when fixing the bug: either keep the number of non-degenerate edges, or keep the number of actual edges. This PR chose the latter, which means that _visually_ the result will change, and so will the point coordinates, but the size of the mesh will be the same. I'd be tempted to think that opting to keep the number of non-degenerate edges (i.e. sticking with a triangle for `resolution=4`) would have the potential to break downstream code relying on the exact size of the polygon (which would arguably be a _good_ thing).
> 
> Anyway, the documentation of the helper clearly states that
> 
> ```
>     resolution : int, default: 100
>         Number of points on the circle.
> ```
> 
> So the current PR's choice is a clear bug-fix that makes the function behave according to the API. Another question is whether this merits a [breaking-change] label... I'm only wondering about this because this will introduce subtle changes with ***no warning*** to end users.

Maybe we can make a warning somehow? But im not sure how to best communicate this change to the end user

GlennWSo commented:
another point i want to make. With this change, pv.Circle will now be behave more like pv.Cylinder
![cylinder4](https://user-images.githubusercontent.com/107837123/207282017-2d402809-31e7-48dc-8f54-2c6b6aa64c90.png)

banesullivan commented:
My opinion: keep the changes as is but add a "breaking change" label to this PR so it shows up in the next release notes. We could add a note in the docstring that prior to version 0.38, this method had incorrect results, which are now fixed. I don't want to throw a warning though

MatthewFlamm commented:
> My opinion: keep the changes as is but add a "breaking change" label to this PR so it shows up in the next release notes. We could add a note in the docstring that prior to version 0.38, this method had incorrect results, which are now fixed. I don't want to throw a warning though

I agree and added the "breaking change" label.  If @GlennWSo can add a section in the PR description describing the breaking change, then I think this meets the ability for users to find this change 'easily'.  Then it is ready to merge IMO.

GlennWSo commented:
> > My opinion: keep the changes as is but add a "breaking change" label to this PR so it shows up in the next release notes. We could add a note in the docstring that prior to version 0.38, this method had incorrect results, which are now fixed. I don't want to throw a warning though
> 
> I agree and added the "breaking change" label.  If @GlennWSo can add a section in the PR description describing the breaking change, then I think this meets the ability for users to find this change 'easily'.  Then it is ready to merge IMO.
> 
> 

Sure, i can do that and include an example.

GlennWSo commented on pyvista/utilities/geometric_objects.py:
```diff
@@ -1288,9 +1288,16 @@ def Circle(radius=0.5, resolution=100):
+
+    Notes
+    --------
+
```
No blank lines allowed between section head and body

dcbr commented:
> another point i want to make. With this change, pv.Circle will now be behave more like pv.Cylinder

This made me look through some other geometric objects and I found 2 others with wrong 'circular resolutions':

<details><summary>CylinderStructured</summary>

```py
pyvista.CylinderStructured(theta_resolution=4).plot()
```
![image](https://user-images.githubusercontent.com/15089458/207549389-2ab7ac02-6f16-4931-9cd1-4b96a600afca.png)

</details>

<details><summary>Ellipse</summary>

```py
pyvista.Ellipse(resolution=4).plot()
```
![image](https://user-images.githubusercontent.com/15089458/207550247-6d86dd74-b0be-41ab-b01c-5424b9e517ef.png)

</details>

I believe the fix for these is the same as for `Circle`, so probably a good idea to include in this PR as well.

GlennWSo commented:
> My opinion: keep the changes as is but add a "breaking change" label to this PR so it shows up in the next release notes. We could add a note in the docstring that prior to version 0.38, this method had incorrect results, which are now fixed. I don't want to throw a warning though

@banesullivan 
## Help wanted
im now struggling to pass the gh action build documentation.
Im not sure if it the rebase i recently made or the changes i made to the docstring that is causing problems.

adeak commented on pyvista/utilities/geometric_objects.py:
```diff
@@ -1288,9 +1288,15 @@ def Circle(radius=0.5, resolution=100):
+    --------
+    prior to version 0.38,
+        this method had incorrect results, which are now fixed.
+    See pull request 3710 for details.
```
1. I would use the `.. versionchanged:` directive for this, assuming that's visible enough. We can check existing examples.
2. Instead of linking to the PR I think we should explain the change in a few words. Most end users don't want to see where the sausage is made :)

banesullivan replied:
Fixed

banesullivan commented:
@GlennWSo, the errors/warnings are burried deep in the build logs. These are what I initially saw that will need to be changed here:

```
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/numpydoc/docscrape.py:460: UserWarning: potentially wrong underline length... 
Notes 
-------- in 
Create a single PolyData circle defined by radius in the XY plane.
... in the docstring of Circle in /home/runner/work/pyvista/pyvista/pyvista/utilities/geometric_objects.py.
  warn(msg)
```

```
/home/runner/work/pyvista/pyvista/pyvista/utilities/geometric_objects.py:docstring of pyvista.utilities.geometric_objects.Circle:24: WARNING: Definition list ends without a blank line; unexpected unindent.
looking for now-outdated files... none found
```
